### PR TITLE
Matrix multiplication and training improvements

### PR DESCRIPTION
**Pull request by Thomas-Chqt**
Hi,
I know this project wasn’t meant to be more than a workshop demo, but I’ve wanted to do something like this for a long time. So, I played around with it a bit and would love to hear your thoughts!

First, I enabled `GIT_SHALLOW` when fetching the Raylib library. I noticed you also did this on your end. As you may have noticed, it doesn’t significantly speed up the fetch time, so I made the git clone command visible in the terminal. Seeing the progress makes it feel faster.

There are other things I would tweak in your CMakeLists.txt, but that’s beyond the scope of this project.

As I mentioned during the workshop, the simplest optimization I can think of is using matrix multiplications in the feedforward function, especially with a library that enables efficient SIMD instructions.

I chose Eigen because it supports arbitrarily sized matrices and is used in TensorFlow.

My implementation of the feedforward function is in `Common/TemplateFeedForward.hpp`. I also added the lander05 executable, which uses the new implementation. Finally, I included tests and benchmarks to verify that my implementation produces the same results and to profile its performance.

The results I got are intriguing:

In debug mode, the matrix implementation is slower than the loop-based one.
```
--------------------------------------------------------------------------------
Benchmark                                             Time       CPU  Iterations
--------------------------------------------------------------------------------
FeedForwardBenchmarck/loop10x3                      142 ns    141 ns     4951861
FeedForwardBenchmarck/matrix10x3                   4176 ns   4155 ns      164806
FeedForwardBenchmarck/loop10x12x12x3                855 ns    851 ns      802053
FeedForwardBenchmarck/matrix10x12x12x3             9273 ns   9241 ns       73765
FeedForwardBenchmarckloop10x20x30x20x10.           3920 ns   3903 ns      176197
FeedForwardBenchmarck/matrix10x20x30x20x10.       15348 ns  15286 ns       44640
FeedForwardBenchmarck/loop10x20x30x40x30x20x10     9388 ns   9339 ns       73371
FeedForwardBenchmarck/matrix10x20x30x40x30x20x10  31636 ns  31474 ns       22060
```

But in release mode, it performs faster.
```
--------------------------------------------------------------------------------
Benchmark                                            Time       CPU   Iterations
--------------------------------------------------------------------------------
FeedForwardBenchmarck/loop10x3                    24.4 ns   24.3 ns     28986347
FeedForwardBenchmarck/matrix10x3                 0.035 ns  0.035 ns  19917938095
FeedForwardBenchmarck/loop10x12x12x3               147 ns    146 ns      4740331
FeedForwardBenchmarck/matrix10x12x12x3            51.0 ns   50.7 ns     14078417
FeedForwardBenchmarck/loop10x20x30x20x10           716 ns    709 ns       997805
FeedForwardBenchmarck/matrix10x20x30x20x10         184 ns    183 ns      3760994
FeedForwardBenchmarck/loop10x20x30x40x30x20x10    1881 ns   1872 ns       372491
FeedForwardBenchmarck/matrix10x20x30x40x30x20x10   396 ns    392 ns      1800569
```

I believe the overhead in debug mode comes from the way I loop through the template arguments using recursive functions. However, since everything is known at compile time, the compiler can fully optimize and flatten the recursion in release mode.

What I don’t understand is why the training iterations aren’t significantly faster with matrix multiplication. Both the training time and the benchmarks show similar performance between the loop-based implementation and the matrix-based implementation.
```
----------------------------------------------------
Benchmark              Time          CPU  Iterations
-----------------------------------------—----------
trainingTask04  76141283 ns  11345938 ns          65
trainingTask05  54687066 ns  11248299 ns          67
```

-----

I also experimented with different parameters to improve the success rate.

First, I tried adding more hidden layers to the network, but it didn’t make a difference.

I noticed that the trained model almost never moves left or right—it only controls its descent and lands when the pad happens to be below the ship. I suspect it gets stuck in a local minimum because staying still results in a more consistent number of high scores.

To address this, I adjusted how the score is calculated. I increased the rewards and penalties for landing, penalized bad landings more than I rewarded good ones, and experimented with switching from an average to a median when calculating the fitness score from the variants. In some cases, I managed to make the model consistently move toward the landing pad, but then it stopped controlling its descent properly.

What do you think could improve the success rate of the trained model?

Also, do you know why the training iterations aren’t accelerated by using matrix multiplication in the feedforward function?

Thank you for your time! I would be happy to hear your thoughts.

### CODE REVIEW

**Comment by dpasca** [comment]
Thanks Thomas, this is great feedback !

The shallow-on is still worth keeping, just in case. I need to remember to do that for all externals, actually.

## Debug and Eigen

First I should say that the version in the workshop was actually building in Debug/unoptimized. That was a mistake, I should have specified Release. I think in the past builds were defaulting to Release/optimized, I'm not sure why not this time, but it's good to keep in mind for the future.

Don't worry about testing the Debug build. Debug builds give a skewed view of performance issues. The plain loop wins simply by the fact that there's no function call to execute a matrix mul, so the compiler can just flatten everything and optimize as much as it can. With unoptimized/Debug builds instead every level of indirection is added cost, because function calls are not optimized out, and there's additional checking code for potential stack corruption.
I see that you're also using new stuff like concepts (I've never seen actual code using them !), which may add extra complexity whose cost may not be fully eliminated for unoptimized builds (just guessing).
Debug builds are a big problem in gamedev, especially when you build your own engine... for this reason I usually have an alternative Debug (DebugOpt) setup with some optimizations enabled, so that the game is still somewhat playable.

Debug builds aside, Eigen can definitely improve things. I used that in a separate project before, and I did get some benefit.

## Other bottlenecks

The reason why you don't see an improvement is probably because the heavy stuff is somewhere else.
Here are a few things to consider, off the top of my head:

1. The simulation itself (low probability). This simulation is actually very simple, so it's probably not the bottleneck, assuming that as much as possible gets inlined in Release.

2. The work done by the training tasks. These can be heavier than the rest, and there's still room for optimization:
- `TrainingTaskGA::evolve()` does not partition work in threads and has a lot of random number generation calls, which may not be optimized for speed. LibTorch (PyTorch) for example has dedicated calls to generate fast noise vectors, because it's so common in this field.
-  `TrainingTaskRES::RunIteration()` does use multi-threading, but only for part of it, and the partition may be too granular (e.g. too many small tasks)

3. The demo's main loop. Currently we execute one training iteration per-frame, and each frame runs at 60Hz (~16.6 ms), this means that even if the training step (`trainingTask.RunIteration()`) takes, let's say 5 ms, the next step training step will only happen after 16.6 ms.
To minimize this bottleneck, it may be a good idea to run multiple training iterations per-frame, like this:

```cpp
for (int i=0; i < 10; ++)
    trainingTask.RunIteration();
```
...this will reduce the dead time due to the vsync by 1/10th
Ideally one simply keeps the training in a separate thread altogether, I do that in [TinyAIDriver](https://github.com/dpasca/TinyAIDriver/blob/427cf3a0a8a0b8aaedf923d389aa056dafce90e1/TinyFreeway/src/TA_TrainingManager.h#L43), but here I wanted to keep things simpler.

## About learning

This is the very tricky part of AI... the fitness/score/loss function is what can make or break everything, but in itself is not sufficient. Here's a short list of things that are all pretty important in the success of training:

### Proper weights of fitness function

As you can see from my revised [CalculateScore()](https://github.com/dpasca/NNLander/blob/619996d9bf327647814780ac6acb8e8ecb0453eb/Common/Simulation.h#L343), it's a delicate balance, and sometimes it's more profitable to ignore some of the elements: currently the fuel consumption is ignored. In an ideal world that would contribute positively, but in practice it can set the training task on a wild goose chase that takes it away from the more important target of landing safely.

### Network size

Bigger is not always better. Again, in an ideal world, a bigger network should be as good or better than a smaller one, after all, if some neurons are not needed, then the training can always set them to 0. However a larger network is hard to train, because there are more variables (the parameters) that need to be guessed correctly.

For this simulation, one hidden layer is probably enough, and it's easier to train successfully.

### Training sample

In the previous version there were not enough training scenarios per-training step. That, coupled with a weak random generator, meant that the landing pad during training was too often to the left and rarely on the right. This mean that the score skewed the learning to do well landing on the left side, but poorly on the right side.

Getting a better custom random generator and increasing the number of training scenarios (SIM_VARIANTS_N) brought much better results. This is a a very simple simulation so increasing the variants further may not be that useful, but it's always something to keep in mind.

### Training approach

The GA approach is rough, but more resilient. In contrast, REINFORCE-ES can work better, but requires more fine-tuning, because of hyperparameters like the learning rate (alpha). This is an issue also with back-propagation, and it's another value that can make or break the training.

### Initial weights condition

This is a mix of technique and pure luck. The initial random values of a network may bring some good development right from the start, in part because of just random luck, in part from the choice of which scale to give to those weights (Xavier vs He vs anything else)

### Pure luck

On the same line of "initial weights condition", there's also the general luck coming from whatever randomness. Here I should also mention that unfortunately the C++ random generators do not guarantee the same results. For this reason, for higher level random generation I usually use a custom random gen... at least for the setup of the simulation, to have the same repeatable environment for every machine.
I'm not doing that for the lower level random use (random weights), so training may result different on different machines or potentially even different version of the standard library...



This is all very interesting stuff.
Regarding the PR, I prefer to keep this repo simple with no special low level optimizations, so that the NN stuff is minimal with no external dependencies.
I'd still like to mention your fork in the readme, as I think it's worth following !

**Comment by Thomas-Chqt** [comment]
@dpasca 

Thank you for your feedback.

I was not really expecting you to merge this PR, it was just a way to add code context to my questions.
My code uses a lot of template meta programming with no comments so I don't think it fits the goal of this project.

About the model training, I know this is a very complicated topic and that there is probably no one fit-for-all solution so it didn't really bother me if it doesn’t work. I also did see the improvement you made on your side in the lander05 and it looks much better.

--------
The most intriguing part was the fact that even if the feed-forward function is faster, the model training speed is not improved.

I did notice that you are only doing one training iteration per frame and that is why the training speed (with the game window) cannot go very fast, and this is why I only benchmarked the `RunIteration` function. So in my case it is not a problem.

To see where the bottleneck is coming from, I tried to profile the app with Xcode's Instruments, but it clearly showed that most of the time is spent in the feed-forward function.
It also reminded me, because of the number of threads in the profiling results, that while you are limiting the number of threads to the number of CPU cores, you are still creating a new thread for each individual and that could cause a bottleneck.

To verify that, I disabled the thread creation completely and reran my benchmarks, and while it is of course slower, the difference between the loop-based implementation and the matrix one is big, and in accordance with the benchmarks of the feed-forward function.

```
---------------------------------------------------------
Benchmark               Time             CPU   Iterations
---------------------------------------------------------
trainingTask04  308712274 ns    307775500 ns            2
trainingTask05   96364340 ns     96077833 ns            6
```
 
What do you think about that, am I missing something ? Can only the thread creation make that much difference ?

If I have time, I will try to make a thread pool to reuse the threads and maybe get the best of the 2 worlds.

Thank you for adding my repo to the readme file. Maybe it will push me to continue this project.

**Comment by dpasca** [comment]
Interesting... for sake of simplicity I used `std::async` which normally is implemented as a pool of threads, so it should be significantly faster than creating an std::thread each time, but actual performance will depend on the implementation of the std lib.
Before the first release, I did some testing to reduce the granularity of threading, but I didn't see any difference, so I didn't investigate further.
It may be worth making your own thread pool, but I don't think it will help too much.

If I get some time I'd like to try profile and play around with how to split in threads. Using OpenMP may be a better solution in some cases.. I haven't used it in a while but it may be better.

Another thing to keep in mind with MT is that it can easily trash the cache.

Another thing is that in an ideal world, the optimizer would work so hard that the FeedForward itself disappears, but after a few nesting levels the optimizer will give up inlining things (some form of "forceinline" may help).

btw here's an old presentation where I talked about SIMD auto-vectorization (about 2/3 into it). I haven't tried in a while though https://newtypekk.com/resources/GDC_2015_FCX_OpenGL_ES_3.1.pdf

**Comment by Thomas-Chqt** [comment]
@dpasca 

I wasn’t aware that `std::async` could use a thread pool. After doing some research, I found that the C++ standard doesn’t mandate the use of a thread pool for `std::async`, and in practice, it seems that only MSVC uses one. That said, I haven’t tested it beyond the NNLander benchmarks.

I don't know if a thread pool would significantly improve performance, but multithreading with modern C++ is an area I’m actively trying to get better at—so it’s always a good topic to explore.

I’ll also take a look at OpenMP—it might offer better performance.

Thanks a lot for your feedback, it’s really helpful!

**Comment by dpasca** [comment]
@Thomas-Chqt 

...ouch, if `std::async` doesn't use a thread pool on Mac/Linux then it's a problem. I guess it's still good enough to use for longer tasks (file reading, etc.), but in this case it could be a substantial bottleneck, unless thread creation is just really fast, but even then, it's probably still considerably slower than keeping a pool of threads.

I think that at some point I'll have to do some benchmarks as well, because there are other projects where I use `std::async` without thinking twice, especially old codebase that was built in Windows.
